src/python/pants/backend/openapi/lint/openapi_format/rules_integration_test.py

```diff
@@ -0,0 +1,153 @@
+from pants.backend.openapi.lint.openapi_format.rules import rules as openapi_format_rules
+from pants.backend.openapi.lint.openapi_format.subsystem import OpenApiFormatFieldSet
+from pants.backend.openapi.target_types import OpenApiSourceGeneratorTarget
+from pants.core.goals.fmt import FmtResult, Partitions
+from pants.core.util_rules import config_files, stripped_source_files
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Address
+from pants.engine.fs import CreateDigest, FileContent, PathGlobs
+from pants.engine.internals.native_engine import Digest, Snapshot
+from pants.engine.target import Target
+from pants.testutil.rule_runner import QueryRule, RuleRunner
+
+
+@pytest.fixture
+def rule_runner() -> RuleRunner:
+    return RuleRunner(
+        rules=[
+            *openapi_format_rules(),
+            *config_files.rules(),
+            *nodejs.rules(),
+            *stripped_source_files.rules(),
+            *target_types_rules(),
+            QueryRule(Partitions, [OpenApiFormatRequest.PartitionRequest]),
+            QueryRule(FmtResult, [OpenApiFormatRequest.Batch]),
+            QueryRule(SourceFiles, (SourceFilesRequest,)),
+        ],
+        target_types=[OpenApiSourceGeneratorTarget],
+    )
+
+
+GOOD_FILE = "openapi: 3.0.0\ninfo:\n  title: Example\n  version: 1.0.0\npaths: {}\n"
+BAD_FILE = "info:\n  title: Example\n  version: 1.0.0\npaths: {}\nopenapi: 3.0.0\n"
+
+
+def run_openapi_format(
+    rule_runner: RuleRunner,
+    targets: list[Target],
+    *,
+    extra_args: list[str] | None = None,
+) -> list[FmtResult]:
+    rule_runner.set_options(
+        [
+            "--backend-packages=pants.backend.experimental.openapi.lint.openapi_format",
+            *(extra_args or ()),
+        ],
+        env_inherit={"PATH"},
+    )
+    field_sets = [OpenApiFormatFieldSet.create(tgt) for tgt in targets]
+    partitions = rule_runner.request(
+        Partitions,
+        [
+            OpenApiFormatRequest.PartitionRequest(tuple(field_sets)),
+        ],
+    )
+
+    results = []
+    for partition in partitions:
+        digest = rule_runner.request(Digest, [PathGlobs(partition.elements)])
+        snapshot = rule_runner.request(Snapshot, [digest])
+        result = rule_runner.request(
+            FmtResult,
+            [
+                OpenApiFormatRequest.Batch(
+                    "",
+                    partition.elements,
+                    partition_metadata=partition.metadata,
+                    snapshot=snapshot,
+                )
+            ],
+        )
+        results.append(result)
+
+    return results
+
+
+def get_snapshot(rule_runner: RuleRunner, source_files: dict[str, str]) -> Snapshot:
+    files = [FileContent(path, content.encode()) for path, content in source_files.items()]
+    digest = rule_runner.request(Digest, [CreateDigest(files)])
+    return rule_runner.request(Snapshot, [digest])
```

Comment:
This will be the 19th declaration of `def get_snapshot(`, so I think it would be worthwhile to have it be a utility function in `testutils` maybe even as a method on `RuleRunner` itself.

Just observing, no need to make that change in this PR (better to make that in a dedicated PR)

Reply:
Taking a quick look, `RuleRunner` has a similar looking method already that works somewhat differently but seemingly the same outcome:
```py
    def make_snapshot(self, files: Mapping[str, str | bytes]) -> Snapshot:
        """Makes a snapshot from a map of file name to file content.

        :API: public
        """
        with temporary_dir() as temp_dir:
            for file_name, content in files.items():
                mode = "wb" if isinstance(content, bytes) else "w"
                safe_file_dump(os.path.join(temp_dir, file_name), content, mode=mode)
            return self.scheduler.capture_snapshots(
                (PathGlobsAndRoot(PathGlobs(("**",)), temp_dir),)
            )[0]
```

Reply:
Yea, that seems like an inferior implementation. Wonder if it bypass some requirement on what rules are in the rule graph?